### DISCUSSION TITLE
Counting/Collecting the CPU samples

### DISCUSSION THREAD

**Discussion by petrbouda** [async-profiler/async-profiler#466] [Q&A] [answered]
Hey guys,

I would ask for some clarification to be sure that I understand correctly how it works :)

Is correct that the sampling is per CPU? If I set up `-i 100ms`
1) 2 CPU -> fully saturated by this process ->  20 samples / 1sec
2) Does it count only the specific process? 
    1 1 CPU runs 
CPU runs 10ms of Process A and 110ms of Process B -> if both had activated monitoring then Process B would throw an interrupt and Process A would have to wait for additional 90ms
3) If I have Process A with 50ms on CPU1 and 60ms on CPU2 then it wouldn't throw an interrupt, am I right?
4) if I set up `-t` then Thread of the Process which crossed the 100ms on the specific CPU (and throws the interrupt) will deliver the interrupt to the handler of the process (implemented in AsyncProfiler) and stores the stacktrace with the specific Thread-name/ID. Am I right? PMU does not distinguish counting between threads of the specific process, it's just about merging the samples inside AsyncProfiler.

Thanks a lot,
Petr

**Comment by apangin** [comment, marked as answer]
Hi Petr,

I assume you mean CPU profiling with perf_events, which is the default mode on Linux. Answers will differ for `itimer` mode.

Async-profiler uses per-thread counters. This allows each thread to handle perf signals generated by itself, and thus combine native stack traces (from perf_events) with Java stack traces (from AsyncGetCallTrace).

1. Most often, yes. In some extreme cases, when an application creates many short-lived threads, the number of samples can be less: e.g. when threads terminate in 50ms, while the sampling interval is 100ms.
2. Yes.
3. No. As stated above, sampling is per-thread. As soon as a thread runs for 100ms in total, no matter on which CPUs, it'll receive a signal.
4. Sampling is always per-thread. `-t` affects only whether samples are grouped by the profiler or not.

Hope this helps.

**Reply by petrbouda** [reply]
Thanks a lot, Andrei!

Yes, I meant `cpu` mode. `itimer` mode is described  here: https://man7.org/linux/man-pages/man2/setitimer.2.html
I noticed it collects significantly fewer samples ... I suppose it's because it counts per Process, not per a single Thread.

If I am correct, theoretically, we would collect the same number of samples in the case of a single-threaded process. 
However, even if we run our application only with `main` thread, JVM itself will start other internal threads of course.

```
ITIMER_PROF
              This timer counts down against the total (i.e., both user
              and system) CPU time consumed by the process.  (The
              measurement includes CPU time consumed by all threads in
              the process.)  At each expiration, a SIGPROF signal is
              generated.

              In conjunction with ITIMER_VIRTUAL, this timer can be used
              to profile user and system CPU time consumed by the
              process.

       A process has only one of each of the three types of timers.

       Timer values are defined by the following structures:

           struct itimerval {
               struct timeval it_interval; /* Interval for periodic timer */
               struct timeval it_value;    /* Time until next expiration */
           };

           struct timeval {
               time_t      tv_sec;         /* seconds */
               suseconds_t tv_usec;        /* microseconds */
           };
```
